base/inference.jl

```diff
@@ -1479,7 +1529,20 @@ function typeinf(linfo::LambdaStaticData,atypes::Tuple,sparams::Tuple, def, cop)
                 end
             end
             pc´ = pc+1
-            if isa(stmt,GotoNode)
+            if isa(changes,StateUpdate) && isa((changes::StateUpdate).var, GenSym)
+                changes = changes::StateUpdate
+                id = (changes.var::GenSym).id+1
+                new = changes.vtype
+                old = gensym_types[id]
+                if old===NF || !(new <: old)
+                    gensym_types[id] = tmerge(old, new)
+                    for r in gensym_uses[id]
+                        if !is(s[r],()) # s[r] === () => unreached statement
```

Comment:
i should probably clarify that comment. we haven't reached it /yet/. since there's only one assignment site for these variables, we consider that all uses of them must have the same type information (using them undef is also disallowed). so we need to make sure to revisit any place that we've already inferred and fix their type information. Using the information that `x::T` after `isa(x,T)` would requiring creating a new GenSym in the generalized algorithm.

but we don't want to use W here to create a control flow branch to a place that may not be reached in normal control flow analysis